### ISSUE TITLE
Compiler allows bad return values from a function when a spread operator is used (version >= 4.1

### ISSUE THREAD

**Issue by kael-shipman** [microsoft/TypeScript#48875]
# Bug Report

### 🔎 Search Terms

* spread operator
* function return

### 🕗 Version & Regression Information

- This changed between versions 4.0 and 4.1

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.1.5#code/PTAEGEHsCdoUwMYBcA2BPUAzFBDA5gM6hywwBQmArgHbICWk1oScBSAjABQ4BcoO1NABpQAIz4C0ASj4BvUIzgB+PtUoBbUSQDczAO6QVoNZpKgAvqFllQoeEkrQm127cV8ARIo9CbFsuZkZCCgAMoAFpCUKAAmWLh4xKTQIqKUSKAxkKzGkEgUNPSMzKxIAEzcEoKpVdJyCtTKqhpa0LpIBkYmrRZWfvaOzn5ujZ7evq6gAHQznKKgAGQLVvqQfGUWUhP+gcFgEVGx8fhJ0DCp6ZnZRNR5BbRIDEwsbADMlfzVYrUyK4pdLR0qwBpmgvRcdjgDicfUmMym3EWy3kt2oAFEAB50NhwahIPhIaCUOCbba2eFzJErW5IABqOBQdBiBKJJPMWz8gXMQA)

### 💻 Code

```ts
// Correctly flags error
function test1(a: any, b: any): { one?: number; two?: number } {
  return {
    one: "one",
  }
}

// Should flag error, but does not
function test2(a: any, b: any): { one?: number; two?: number } {
  return {
    one: "one",
    ...(b && { two: 2 }),
  }
}

// Should flag error, but does not
function test3(a: any, b: any): { one?: number; two?: number } {
  return {
    ...(a && { nonExistent: true }),
    ...(b && { notValid: true }),
  }
}
```

### 🙁 Actual behavior

Typescript does not flag the erroneous keys and/or values being returned

### 🙂 Expected behavior

Typescript should flag the erroneous keys and/or values being returned (it does this correctly in versions prior to 4.1)

**Comment by andrewbranch** [comment]
The unsoundness only happens because your `a` and `b` are `any`. Prior to 4.1, there was a bug in `&&` binary expressions that assumed `any` was always truthy, so the widened type of `a && { nonExistent: true }` was `{ nonExistent: boolean }`. But `any` can include all manner of falsey values, so this was incorrect. The most accurate type for the expression would be the union of all falsey types and `{ nonExistent: boolean }`, but a large union like that is expensive and is inconsistent with the way `any` infects other types (`any & T` is `any`), so we fixed the type of `a && { nonExistent: true }` to be `any`. So consequently the type of your return expression in all these is just `any`, which is not perfect, but that is exactly what you sign up for when you mix `any` into stuff.

**Comment by andrewbranch** [comment]
`unknown` will work just as well for your callers; it just might be a bit of a pain to work with internally as it doesn’t narrow under most circumstances.

**Comment by kael-shipman** [comment]
The real function I was writing had a few in-line checks in it, such as

```ts
if (a && a.something) {
  // do something
}
if (b && typeof b.somethingElse === "string") {
  // do something else
}
```

That's why `any` was more convenient in this case. Unless maybe there's some other paradigm for those kinds of things that I haven't seen yet? (Note that for most runtime validations I use `runtypes`. In this case, I also used `runtypes`, but there was some additional stuff I had to do that required some pre-arranging/checking.)
